csharp/src/Apache.Arrow/DecimalUtility.cs

```diff
@@ -0,0 +1,144 @@
+    /// </summary>
+    public static class DecimalUtility
+    {
+        private static readonly BigInteger _maxDecimal = new BigInteger(decimal.MaxValue);
+        private static readonly BigInteger _minDecimal = new BigInteger(decimal.MinValue);
+        private static readonly ulong[] s_powersOfTen =
+        {
+            1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000, 10000000000, 100000000000,
+            1000000000000, 10000000000000, 100000000000000, 1000000000000000, 10000000000000000, 100000000000000000,
+            1000000000000000000, 10000000000000000000
+        };
+
+        private static int PowersOfTenLength => s_powersOfTen.Length - 1;
+
+        public static decimal GetDecimal(in ArrowBuffer valueBuffer, int index, int scale, int byteWidth,
+            bool isUnsigned = false)
+        {
+            int startIndex = index * byteWidth;
+            ReadOnlySpan<byte> value = valueBuffer.Span.Slice(startIndex, byteWidth);
+            BigInteger integerValue;
+
+#if NETCOREAPP
+            integerValue = new BigInteger(value);
+#else
+            integerValue = new BigInteger(value.ToArray());
+#endif
+
+            if (integerValue > _maxDecimal || integerValue < _minDecimal)
+            {
+                BigInteger scaleBy = BigInteger.Pow(10, scale);
+                BigInteger integerPart = BigInteger.DivRem(integerValue, scaleBy, out BigInteger fractionalPart);
+                if (integerPart > _maxDecimal || integerPart < _minDecimal) // decimal overflow, not much we can do here - C# needs a BigDecimal
+                {
+                    throw new OverflowException("Value: " + integerPart + " too big or too small to be represented as a decimal");
+                }
+                return (decimal)integerPart + DivideByScale(fractionalPart, scale);
+            }
+            else
+            {
+                return DivideByScale(integerValue, scale);
+            }
+        }
+
+        private static decimal DivideByScale(BigInteger integerValue, int scale)
+        {
+            decimal result = (decimal)integerValue; // this cast is safe here
+            int drop = scale;
+            while (drop > PowersOfTenLength)
+            {
+                result /= s_powersOfTen[PowersOfTenLength];
+                drop -= PowersOfTenLength;
+            }
+
+            result /= s_powersOfTen[drop];
+            return result;
+        }
+
+        public static void GetBytes(decimal value, int precision, int scale, int byteWidth, Span<byte> bytes)
+        {
+            // create BigInteger from decimal
+            byte[] bigIntBytes = new byte[12];
+            int[] decimalBits = decimal.GetBits(value);
+            int decScale = (decimalBits[3] >> 16) & 0x7F;
+            for (int i = 0; i < 3; i++)
+            {
+                int bit = decimalBits[i];
+                byte[] intBytes = BitConverter.GetBytes(bit);
```

Comment:
Should be able to call `BitConverter.TryWriteBytes(Span<byte>, int)` on NETCOREAPP to get rid of this allocation.